NOnion/Directory/NetworkStatusDocument.fs

```diff
@@ -0,0 +1,303 @@
+            Identity = None
+            Address = None
+            IP = None
+            DirectoryPort = None
+            OnionRouterPort = None
+            Contact = None
+            VoteDigest = None
+        }
+
+type RouterStatusEntry =
+    {
+        NickName: Option<string>
+        Identity: Option<string>
+        Digest: Option<string>
+        PublicationTime: Option<DateTime>
+        IP: Option<string>
+        OnionRouterPort: Option<int>
+        DirectoryPort: Option<int>
+    }
+
+    static member Empty =
+        {
+            RouterStatusEntry.NickName = None
+            Identity = None
+            Digest = None
+            PublicationTime = None
+            IP = None
+            DirectoryPort = None
+            OnionRouterPort = None
+        }
+
+type DirectorySignature =
+    {
+        Identity: Option<string>
+        Digest: Option<string>
+        Signature: Option<string>
+    }
+
+    static member Empty =
+        {
+            DirectorySignature.Identity = None
+            Digest = None
+            Signature = None
+        }
+
+    static member Parse (lines: Queue<string>) =
+        let rec innerParse state =
+            let rec readMultiLineUntilEnd (state: string) =
+                let line = lines.Dequeue ()
+
+                if line <> "-----END SIGNATURE-----" then
+                    readMultiLineUntilEnd (state + line)
+                else
+                    state
+
+            let words = lines.Dequeue().Split ' ' |> Queue<string>
+
+            let readWord () =
+                words.Dequeue ()
+
+            match words.Dequeue () with
+            | "directory-signature" ->
+                innerParse
+                    { state with
+                        DirectorySignature.Identity = readWord () |> Some
+                        Digest = readWord () |> Some
+                    }
+            | "-----BEGIN" ->
+                { state with
+                    Signature = readMultiLineUntilEnd String.Empty |> Some
+                }
+            | _ -> failwith "failed to parse signature"
+
+        innerParse DirectorySignature.Empty
+
+type NetworkStatusDocument =
+    {
+        Version: Option<int>
+        VoteStatus: Option<string>
+        ConsensusMethod: Option<int>
+        ValidAfter: Option<DateTime>
+        FreshUntil: Option<DateTime>
+        ValidUntil: Option<DateTime>
+        VotingDelay: Option<string>
+        ClientVersions: Option<string>
+        ServerVersions: Option<string>
+        KnownFlags: Option<string>
+        RecommendedClientProtocols: Option<string>
+        RecommendedRelayProtocols: Option<string>
+        RequiredClientProtocols: Option<string>
+        RequiredRelayProtocols: Option<string>
+        Params: Option<string>
+        SharedRandomPreviousValue: Option<string>
+        SharedRandomCurrentValue: Option<string>
+        BandwithWeights: Option<string>
+
+        Packages: list<string>
```

Comment:
use List, not list; lowercase is for primitive types

Reply:
still not addressed: use List here and SCG.Queue for the Queue

Reply:
Done